### PR TITLE
ddtrace/tracer: add horizontal propagation for sampling mechanism

### PR DESCRIPTION
**Pull request by ajgajg1134**
This will forward any (valid) incoming tags in `x-datadog-tags` and if _dd.p.dm is not set then it will be set based on the sampling decision made by this service.

EDIT: Sending back to draft to await clarification on comments from PR
EDIT2: back to ready for review :)

### CODE REVIEW

File: internal/servicehash/servicehash.go
Comment (katiehockman) [inline internal/servicehash/servicehash.go:4]: Copyright 2022?

File: internal/servicehash/servicehash.go
Comment (katiehockman) [inline internal/servicehash/servicehash.go:15]: Just curious, where did this cache size come from?

File: internal/servicehash/servicehash.go
Comment (katiehockman) [inline internal/servicehash/servicehash.go:31]: Comment says "first 10 characters" but the code is `[:5]`. Is one of these incorrect or am I misunderstanding this?

**Review by katiehockman — COMMENTED** [review]
Is there an issue describing the context around this change? Thanks!

File: internal/servicehash/servicehash.go
Comment (ajgajg1134) [reply internal/servicehash/servicehash.go:31]: 👍 I was worried this comment wasn't quite clear enough :P Yeah since `hashb` is a byte array each byte is expanded to two hex characters in the string representation. How about adding something like  `2 chars per byte * 5 bytes = 10 chars` to the comment to make it more clear?

File: ddtrace/tracer/spancontext.go
Comment (ajgajg1134) [reply ddtrace/tracer/spancontext.go:252]: Yep! If the `propagateServiceName` configuration is either unavailable or is disabled then the service hash that is propagated to the next service will be the empty string. The `-` and the priority number can remain though. There's some additional details about this in the RFC

File: internal/servicehash/servicehash.go
Comment (ajgajg1134) [reply internal/servicehash/servicehash.go:15]: Good question, I got this file from the original work that Barbayar was doing. @Barbayar any info on why you picked this cache size? Or did it just seem like a reasonable default / size?

File: internal/servicehash/servicehash.go
Comment (katiehockman) [reply internal/servicehash/servicehash.go:31]: Yeah that's more descriptive! 👍

File: internal/servicehash/servicehash.go
Comment (Barbayar) [reply internal/servicehash/servicehash.go:15]: There was no particular reason for choosing `1000`. You can adjust it if you want.

File: ddtrace/tracer/option.go
Comment (Barbayar) [reply ddtrace/tracer/option.go:43]: The spec has changed. The main motivation is to avoid `HTTP Status 413`.

File: ddtrace/tracer/option_test.go
Comment (gbbr) [inline ddtrace/tracer/option_test.go:425]:
```suggestion
		})

		t.Run("default", func(t *testing.T) {
```

File: ddtrace/tracer/option.go
Comment (gbbr) [inline ddtrace/tracer/option.go:273]:
```suggestion
		envKey := "DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH"
		max := internal.IntEnv(envKey, defaultMaxTagsHeaderLen)
		if max < 0 {
			log.Warn("Invalid value %d for %s. Setting to 0.", max, envKey)
			max = 0
		}
		if max > 512 {
			log.Warn("Invalid value %d for %s. Maximum allowed is %d. Setting to %d.", max, envKey, 512, 512)
			max = 512
		}
```
A few things:
* I don't think there's any need for a separate `clamp` function. It's easier to inline things.
* Numbers like `512` should be constants.
* If the value of the env. var. is not a number, I think the user should know. The way `internal.IntEnv` is implemented is not that great. If the env. var. is not set, then it makes sense to use the default, but if the env. var. is set and the value is invalid (e.g. not a number), I think the user should be warned.

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [inline ddtrace/tracer/spancontext.go:245]:
```suggestion
// setServiceDecisionMaker sets the service and sampler which took the sampling decision for this trace and span.
```

File: internal/env.go
Comment (gbbr) [inline internal/env.go:30]: I think this is a good change. Thanks for making it. Do you think we should do it for `BoolEnv` too?

**Review by gbbr — APPROVED** [review]

File: internal/env.go
Comment (gbbr) [inline internal/env.go:30]: Oh, wait. Actually, this will return an error even if the env. var. is not set. So careful here, it could turn into a regression. If the env. var. is not set, we should return a default, otherwise, log an error + the default return.
```go
vv, ok := os.LookupEnv(key)
if !ok {
    return def
}
v, err := strconv.Atoi(vv)
// ...
```

**Review by gbbr — CHANGES_REQUESTED** [review]

File: ddtrace/tracer/option.go
Comment (ajgajg1134) [reply ddtrace/tracer/option.go]: 👍 good call out, I've changed the name and updated the comment so hopefully it's more clear (and not just self-referential 😛 )

File: ddtrace/tracer/option.go
Comment (gbbr) [inline ddtrace/tracer/option.go:542]:
```suggestion
// propagated to other services (e.g. over HTTP requests).
```
What feature does this propagation enable? Why would a user want to set this option?

Comment (ajgajg1134) [reply ddtrace/tracer/option.go]: 👍 Added a note about the new ingestion control page, let me know if you think that makes it clear enough 😃

File: ddtrace/tracer/option.go
Comment (gbbr) [inline ddtrace/tracer/option.go:182]:
```suggestion
// maxPropagatedTagsLength limits the size of DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH to prevent HTTP 413 responses.
```

File: ddtrace/tracer/option.go
Comment (gbbr) [inline ddtrace/tracer/option.go:543]:
```suggestion
// propagated to other services (e.g. over HTTP requests). This enhances the ingestion control page
// so that it can show which exact service is the root of ingested spans, and therefore cost.
```
nit: it won't be new after some time passes, and then the docs will just be weird :)

Thank you so much for adding this text. This makes such a big difference to a user! Otherwise a "horizontal service hash propagation" means nothing. Disclaimer: I didn't know either what it meant until this moment!

nit 2: in order to have a prettier and shorter API, does it make sense to call it `WithServicePropagation`? Reason being that we've also pretty much gone with `Service` instead of `ServiceName` (e.g. see `WithService` vs `WithServiceName` which is the older version of it).

File: internal/servicehash/servicehash.go
Comment (gbbr) [inline internal/servicehash/servicehash.go:14]: You are using the grouping notation here (`(`) for just one constant, but elsewhere in the code you are using no grouping for multiple constants. Seems a bit counter intuitive.

NIT!

File: internal/servicehash/servicehash.go
Comment (gbbr) [inline internal/servicehash/servicehash.go:6]: Since this is only used in the `ddtrace/tracer` package, why did it warrant a separate package?

Comment (Barbayar) [reply internal/servicehash/servicehash.go:6]: This part was written by me.
@ajgajg1134  Feel free to update the package name.

Comment (gbbr) [reply internal/servicehash/servicehash.go:6]: I just meant it can be put inside `ddtrace/tracer` and doesn't need a separate package because nobody else is using it.

Comment (ajgajg1134) [reply internal/servicehash/servicehash.go]: 👍 I've moved it into the ddtrace/tracer package. If we ever want to use it elsewhere we can always just move it back out to a shared place

File: ddtrace/tracer/textmap.go
Comment (katiehockman) [inline ddtrace/tracer/textmap.go:123]:
```suggestion
	// It defaults to defaultMaxTagsHeaderLen, a value of 0 disables propagation of tags.
```

Comment (katiehockman) [inline ddtrace/tracer/textmap.go:123]:
```suggestion
	// It defaults to defaultMaxTagsHeaderLen, a value of 0 disables propagation of tags.
```

**Review by katiehockman — APPROVED** [review]
Nice, thanks Andrew! Changes look good. Probably get an additional approval from someone with more context too before merging 👍

File: ddtrace/tracer/option.go
Comment (gbbr) [inline ddtrace/tracer/option.go:272]: Since this env. key is used in multiple places, should we make it into a constant? Granted we should do that with all of them...

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [inline ddtrace/tracer/servicehash.go:21]:
```suggestion
// Hash computes the hash for `service`. It will use a local cache to avoid recomputation.
```

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [inline ddtrace/tracer/servicehash.go:34]: Can we drop the `get*` prefix? (following https://go.dev/doc/effective_go#Getters)

Also, it would be more elegant (and standard) to return a second argument here.

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [inline ddtrace/tracer/servicehash.go:26]:
```suggestion
	if h, ok := hashFromCache(service); ok {
		return h
	}
```
I don't know if there's any point to go through the trouble of re-using this `hash` variable.

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [inline ddtrace/tracer/servicehash.go:28]: This is confusing, you're slicing at 5 below

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [inline ddtrace/tracer/servicehash.go:47]: The efficiency of it will be random too, once it gets full. Not sure if the edge case is worth it, but I'd implement a proper LRU cache, something like: https://github.com/golang/go/blob/2882786bf4cd779f166e9ced82a4da2ea0f8b1f9/src/net/http/transport.go#L2868-L2871

I don't feel strongly about this, so if you can't afford the time or effort, I'm ok with merging as is since the gain of this is very on the edge case side.

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [inline ddtrace/tracer/spancontext.go:237]: Can we spell. out what's going on here? i.e.
```suggestion
		// we have a positive priority and the decision maker isn't set
		t.setServiceDecisionMaker(service, sampler, span)
```

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [inline ddtrace/tracer/spancontext.go:240]:
```suggestion
		// a negative priority removes a decision maker
		delete(t.propagatingTags, keyDecisionMaker)
```
Why are we not setting the decision maker for `p <= 0`?

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [inline ddtrace/tracer/spancontext.go:253]: What happens if there is no tracer above? Will this tag have the value starting with `-sampler_name` surely that doesn't look right. We should have a default value, like "unknown" or something.

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [inline ddtrace/tracer/spancontext.go:248]:
```suggestion
	tr, ok := internal.GetGlobalTracer().(*tracer)
```
I think there aren't any other `ok` values in this small function to warrant a special non-standard name.

Also, I suspect, based on the answer to the previous question, that we may be able to simply return if `!ok` or add a default value.

File: ddtrace/tracer/option.go
Comment (ajgajg1134) [reply ddtrace/tracer/option.go:238]: It changed due to the RFC getting updated based on feedback from the Product folks that the default should be true

File: ddtrace/tracer/spancontext.go
Comment (ajgajg1134) [reply ddtrace/tracer/spancontext.go:254]: If there is no tracer (or the service name propagation is disabled) The RFC calls for this, admittedly odd, format.

File: ddtrace/tracer/option.go
Comment (ajgajg1134) [reply ddtrace/tracer/option.go:272]: This env key isn't actually used anywhere else as far as my searching can see? Am I missing a spot here? 

(The reason I didn't make it a constant to begin with was it only exists in this one spot afaik)

File: ddtrace/tracer/servicehash.go
Comment (ajgajg1134) [reply ddtrace/tracer/servicehash.go:28]: Hmm any ideas on how to make this clearer? I added the comment to clarify that we want 10 characters of hash and since each byte becomes 2 hex characters we need to slice only the first 5 bytes of hash data to end up with a 10 character long hex string

File: ddtrace/tracer/servicehash.go
Comment (ajgajg1134) [reply ddtrace/tracer/servicehash.go:47]: Yeah this is actually a question I have, it seems like the vast majority of go programs wouldn't have more than 1k unique service names? Or are there typical customer use-cases that could result in filling this cache regularly?

(Personally I like the simplicity here by avoiding LRU logic, but you are definitely right that when the cache is full the performance will suffer)

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [reply ddtrace/tracer/spancontext.go:241]: Right, but why don't we need a decision maker if p<=0. That's a decision too, and it will reach the agent. Or is this decision maker only used by the backend? In which case it makes sense, and we should comment that.

File: ddtrace/tracer/option.go
Comment (gbbr) [reply ddtrace/tracer/option.go:272]: You're right! My bad!

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [reply ddtrace/tracer/spancontext.go:249]: You forgot this one.

File: ddtrace/tracer/spancontext.go
Comment (ajgajg1134) [reply ddtrace/tracer/spancontext.go:241]: The decision maker is primarily for the backend so that it can highlight _why_ a particular span was kept. This is mostly in support of the a better ingestion control page so customers can better see where their spans are coming from and change the sampling rules in those services if desired. @Barbayar - feel free to correct me here or provide some additional context if needed!

Comment (Barbayar) [reply ddtrace/tracer/spancontext.go:241]: Yes, it's used by only on backend. There are two reasons.
- Highlighting the span made the decision on the UI, so customer knows that which span caused the sampling.
- Correlate hash and service. We can build the hash table on the backend from services we've seen. But, The hash tracers compute are based on unnormalized service names, but the hashes backend computes on the backend will be based on normalized service names. So, we will build a fallback table using this tag.

**Review by katiehockman — APPROVED** [review]

File: ddtrace/tracer/option.go
Comment (ajgajg1134) [reply ddtrace/tracer/option.go]: It changed as a result of a review of the RFC. As I understand it, the default behavior will be to have this feature enabled so customers can use the new ingestion control page without having to do a lot of configuration across all their different services. (Also it looks like you're reviewing an outdated view of the code? The name of the is env var has also changed as a result of trying to be more clear / concise and was voted on in a poll)

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [reply ddtrace/tracer/servicehash.go:27]: It's clear now. Thanks 🙂

File: ddtrace/tracer/servicehash.go
Comment (gbbr) [reply ddtrace/tracer/servicehash.go:46]: Yeah, given the amount of times this may get called, I think that optimising here might be overkill. Resolving.

File: ddtrace/tracer/spancontext.go
Comment (gbbr) [reply ddtrace/tracer/spancontext.go:241]: Thanks for explaining 🙇🏻

File: ddtrace/tracer/textmap_test.go
Comment (gbbr) [inline ddtrace/tracer/textmap_test.go:184]: How come you didn't use sub-tests?

File: ddtrace/tracer/tracer_test.go
Comment (gbbr) [inline ddtrace/tracer/tracer_test.go:22]: I think that the previous order was the correct one.

**Review by gbbr — APPROVED** [review]
LGTM!

**Comment by ajgajg1134** [comment]
Ready for review and then merge after approval! :) 
One important change: `DD_TRACE_TAGS_PROPAGATION_MAX_LENGTH` was added as part of the experimental sql commenting, however it wasn't actually ever used anywhere so far as I can tell. The RFC for Trace Tag propagation uses `DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH` instead so that's what I've replaced it with. CC @alexandre-normand as a heads-up

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:282]: I think the name is not quite right. This isn't formatting anything per se, it's creating the tags. So maybe `marshalPropagatingTags` or something like that. Also, why return the builder? Why not just `string`?

Please document this function.

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:291]:
```suggestion
			log.Warn("Won't propagate tag '%s': %v)", k, err.Error())
```

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:290]:
```suggestion
		if err := isValidPropagatableTag(k, v); err != nil {
```

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:297]:
```suggestion
			log.Warn("Won't propagate trace tags: maximum trace tags header length (%d) reached.", p.cfg.MaxTagsHeaderLen)
```
Please remember that this is for humans. Log messages should be capitalised.

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:362]: Please document this function. What does it do? What does it extract? It has no return value.

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:369]:
```suggestion
		log.Warn("Did not extract trace tags, size limit exceeded: %d", propagationExtractMaxSize)
```
What is the actual impact on the user here? Should "trace tags" mean anything to them? This seems more like an internal term. If there is an actual impact, let's refer specifically to that. If there is no impact on the user, then let's not warn anything.

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:375]:
```suggestion
		log.Warn("Did not extract trace tags: %v.", err.Error())
```

File: ddtrace/tracer/textmap_test.go
Comment (gbbr) [inline ddtrace/tracer/textmap_test.go:292]: Why aren't these just one test with sub-tests? Their naming is clearly hinting at it.

**Review by gbbr — DISMISSED** [review]

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:99]: Can we group them with the rest of the bunch? (and maybe use the same naming style)

https://github.com/DataDog/dd-trace-go/blob/bd20964575c07f00d616972129c540f47b5dcd81/ddtrace/tracer/span.go#L563-L573

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:276]:
```suggestion
	if s := p.marshalPropagatingTags(ctx); len(s) > 0 {
```
The `if` improvement is just a nit. Feel free to ignore. The comment on the other hand can be removed. It's clearly visible from the code that `s` gets written.

File: ddtrace/tracer/textmap.go
Comment (gbbr) [inline ddtrace/tracer/textmap.go:371]: 👍🏻 Very nice. Thanks for adding the context. Now it's clear what the impact is.

File: ddtrace/tracer/textmap_test.go
Comment (gbbr) [inline ddtrace/tracer/textmap_test.go:273]:
```suggestion
	})

	t.Run("ExtractTraceTagsTooLong", func(t *testing.T) {
```

**Pull request merged**
